librosa/util/utils.py

```diff
@@ -35,80 +35,168 @@
+
+    This generalizes to higher dimensional inputs, as shown in the examples below.
+    In general, the framing operation increments by 1 the number of dimensions,
+    adding a new "frame axis" either to the end of the array (`axis=-1`)
+    or the beginning of the array (`axis=0`).
+
 
     Parameters
     ----------
-    y : np.ndarray [shape=(n,)]
-        Time series to frame. Must be one-dimensional and contiguous
-        in memory.
+    x : np.ndarray
+        Time series to frame. Must be contiguous in memory, see the "Raises"
+        section below for more information.
 
     frame_length : int > 0 [scalar]
-        Length of the frame in samples
+        Length of the frame
 
     hop_length : int > 0 [scalar]
-        Number of samples to hop between frames
+        Number of steps to advance between frames
+
+    axis : 0 or -1
+        The axis along which to frame.
+
+        If `axis=-1` (the default), then `x` is framed along its last dimension.
+        `x` must be "F-contiguous" in this case.
+
+        If `axis=0`, then `x` is framed along its first dimension.
+        `x` must be "C-contiguous" in this case.
 
     Returns
     -------
-    y_frames : np.ndarray [shape=(frame_length, N_FRAMES)]
-        An array of frames sampled from `y`:
-        `y_frames[i, j] == y[j * hop_length + i]`
+    x_frames : np.ndarray [shape=(..., frame_length, N_FRAMES) or (N_FRAMES, frame_length, ...)]
+        A framed view of `x`, for example with `axis=-1` (framing on the last dimension):
+        `x_frames[..., j] == x[..., j * hop_length : j * hop_length + frame_length]`
+
+        If `axis=0` (framing on the first dimension), then:
+        `x_frames[j] = x[j * hop_length : j * hop_length + frame_length]`
 
     Raises
     ------
     ParameterError
-        If `y` is not contiguous in memory, not an `np.ndarray`, or
-        not one-dimensional.  See `np.ascontiguous()` for details.
+        If `x` is not contiguous in memory or not an `np.ndarray`.
+
+        If `x.shape[axis] < frame_length`, there is not enough data to fill one frame.
 
         If `hop_length < 1`, frames cannot advance.
 
-        If `len(y) < frame_length`.
+        If `axis` is not 0 or -1.  Framing is only supported along the first or last axis.
+            If `axis=-1` (the default), then `x` must be "F-contiguous".
+            If `axis=0`, then `x` must be "C-contiguous".
+
+        If the contiguity of `x` is incompatible with the framing axis.
+
+    See Also
+    --------
+    np.asfortranarray : Convert data to F-contiguous representation
+    np.ascontiguousarray : Convert data to C-contiguous representation
+    np.ndarray.flags : information about the memory layout of a numpy `ndarray`.
 
     Examples
     --------
-    Extract 2048-sample frames from `y` with a hop of 64 samples per frame
+    Extract 2048-sample frames from monophonic `y` with a hop of 64 samples per frame
```

Comment:
is that a realistic use case? 97% overlap?

Reply:
3ms hop gets used in onset detection, yeah